myia/ir.py

```diff
@@ -1,7 +1,222 @@
+            special value `PARAMETER`.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        uses: A set of tuples with the nodes that use this node alongside with
+            the index. These def-use edges are the reverse of the `inputs`
+            attribute, creating a doubly linked graph structure.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self, inputs: List['Node'], value: Any,
+                 graph: Optional[Graph]) -> None:
+        """Construct a node."""
+        self.inputs = Inputs(self, inputs)
+        self.value = value
+        self.graph = graph
+        self.uses: Set[Tuple[Node, int]] = set()
+        self.debug: Dict = {}
+
+
+class Inputs(MutableSequence[Node]):
+    """Container data structure for node inputs.
+
+    This mutable sequence data structure can be used to keep track of a node's
+    inputs. Any insertion or deletion of edges will be reflected in the inputs'
+    `uses` attribute.
+
+    """
+
+    def __init__(self, node: Node, initlist: Iterable[Node] = None) -> None:
+        """Construct the inputs container for a node.
+
+        Args:
+            node: The node of which the inputs are stored.
+            initlist: A sequence of nodes to initialize the container with.
+
+        """
+        self.node = node
+        self.data: List[Node] = []
+        if initlist is not None:
+            self.extend(initlist)
+
+    @overload
+    def __getitem__(self, index: int) -> Node:
+        pass
+
+    @overload  # noqa: F811
+    def __getitem__(self, s: slice) -> Sequence[Node]:
+        pass
+
+    def __getitem__(self, index):  # noqa: F811
+        """Get an input by its index."""
+        return self.data.__getitem__(index)
+
+    @overload
+    def __setitem__(self, index: int, value: Node) -> None:
+        pass
+
+    @overload  # noqa: F811
+    def __setitem__(self, index: slice, value: Iterable[Node]) -> None:
+        pass
+
+    def __setitem__(self, index, value):  # noqa: F811
+        """Replace an input with another."""
+        if isinstance(index, slice):
+            for i, v in zip(range(*index.indices(len(self))), value):
+                self[i] = v
+        old_value = self.data[index]
+        if old_value.uses is not None:
+            old_value.uses.remove((self.node, index))
+        if value.uses is not None:
+            value.uses.add((self.node, index))
+        self.data.__setitem__(index, value)
+
+    @overload
+    def __delitem__(self, index: int) -> None:
+        pass
+
+    @overload  # noqa: F811
+    def __delitem__(self, index: slice) -> None:
+        pass
+
+    def __delitem__(self, index):  # noqa: F811
+        """Delete an input."""
+        if isinstance(index, slice):
+            for i in range(*index.indices(len(self))):
+                del self[i]
```

Comment:
This will not delete what the users expects.

Reply:
I actually think there's a bunch of bugs in there, but I wanted to push the general concept before going through more thorough testing.